src/ci/pgo-windows.sh

```diff
@@ -0,0 +1,173 @@
+#!/bin/bash
+# ignore-tidy-linelength
+
+set -euxo pipefail
+
+ci_dir=`cd $(dirname $0) && pwd`
+source "$ci_dir/shared.sh"
+
+CHECKOUT=$(pwd)
+
+DOWNLOADED_LLVM=$CHECKOUT/citools/clang-rust
+
+BUILD=$CHECKOUT/build/$PGO_HOST
+RUSTC_STAGE_0=$BUILD/stage0/bin/rustc.exe
+CARGO_STAGE_0=$BUILD/stage0/bin/cargo.exe
+RUSTC_STAGE_2=$BUILD/stage2/bin/rustc.exe
+
+# Make sure we have a temporary PGO work folder
+PGO_TMP=$CHECKOUT/build/tmp-pgo
+mkdir -p $PGO_TMP
+rm -rf $PGO_TMP/*
+
+RUSTC_PERF=$PGO_TMP/rustc-perf
+
+# Compile several crates to gather execution PGO profiles.
+# Arg0 => profiles (Debug, Opt)
+# Arg1 => scenarios (Full, IncrFull, All)
+# Arg2 => crates (syn, cargo, ...)
+gather_profiles () {
+  cd $CHECKOUT
+
+  # Compile libcore, both in opt-level=0 and opt-level=3
+  RUSTC_BOOTSTRAP=1 $RUSTC_STAGE_2 \
+      --edition=2021 --crate-type=lib ./library/core/src/lib.rs \
+      --out-dir $PGO_TMP
+
+  RUSTC_BOOTSTRAP=1 $RUSTC_STAGE_2 \
+      --edition=2021 --crate-type=lib -Copt-level=3 ./library/core/src/lib.rs \
+      --out-dir $PGO_TMP
+
+  cd $RUSTC_PERF
+
+  # Run rustc-perf benchmarks
+  # Benchmark using profile_local with eprintln, which essentially just means
+  # don't actually benchmark -- just make sure we run rustc a bunch of times.
+  RUST_LOG=collector=debug \
+  RUSTC=$RUSTC_STAGE_0 \
+  RUSTC_BOOTSTRAP=1 \
+  $CARGO_STAGE_0 run -p collector --bin collector -- \
+      profile_local \
+      eprintln \
+      $RUSTC_STAGE_2 \
+      --id Test \
+      --profiles $1 \
+      --cargo $CARGO_STAGE_0 \
+      --scenarios $2 \
+      --include $3
+
+  cd $CHECKOUT
+}
+
+# This path has to be absolute
+LLVM_PROFILE_DIRECTORY_ROOT=$PGO_TMP/llvm-pgo
+
+# We collect LLVM profiling information and rustc profiling information in
+# separate phases. This increases build time -- though not by a huge amount --
+# but prevents any problems from arising due to different profiling runtimes
+# being simultaneously linked in.
+# LLVM IR PGO does not respect LLVM_PROFILE_FILE, so we have to set the profiling file
+# path through our custom environment variable. We include the PID in the directory path
+# to avoid updates to profile files being lost because of race conditions.
+# On windows, when building LLVM for PGO, for some reason, the profiler runtime library is not
+# linked in automatically so we do it manually.
+clang_builtins=$($DOWNLOADED_LLVM/bin/clang-cl /clang:-print-libgcc-file-name /clang:--rtlib=compiler-rt)
+profiler_rt_path=$(dirname $clang_builtins)
+RUSTFLAGS="-Clink-arg=/LIBPATH:${profiler_rt_path}" \
+LLVM_PROFILE_DIR=${LLVM_PROFILE_DIRECTORY_ROOT}/prof-%p python ./x.py build \
+    --target=$PGO_HOST \
+    --host=$PGO_HOST \
+    --stage 2 library/std \
+    --llvm-profile-generate
```

Comment:
@michaelwoerister this is the most incomprehensible thing I've had to work around (just to get something working, I don't propose we land this as-is).

Similarly to building lld needing the profiler runtime, if I don't do that here, linker errors also happen while linking rustc-driver ([example output of the error](https://gist.github.com/lqd/258a608aba66c325c00e6c3382542bfa#file-linker-error-txt)) but not while building LLVM, do you think it's something we can fix by changing some cmake flags ? (seems unlikely, since LLVM was built with the flags we see in their cmake setup, [for example, see here](https://gist.github.com/lqd/258a608aba66c325c00e6c3382542bfa#file-llvm-cmake-txt))

Reply:
Also cc @Mark-Simulacrum since you've worked on landing the linux PGO initially, that may ring a bell (although it seems very specific to `clang-cl`), and have a better idea on where such a linker argument change could be made more cleanll within bootstrap.

Reply:
I'd loosely expect it to be the compiler's responsibility to pass that linker argument, not bootstrap. Are we doing something unusual that makes that not work?

Reply:
I don't believe so, but it seems specific to building things with clang-cl with instrumentation, LLVM's cmake has [something similar](https://github.com/llvm/llvm-project/blob/782e91224601e461c019e0a4573bbccc6094fbcd/llvm/cmake/modules/HandleLLVMOptions.cmake#L1058-L1079).